eval.py

```diff
@@ -0,0 +1,203 @@
+import os
+import json
+import pickle
+import hashlib
+
+import hydra
+import torch
+import torch.nn as nn
+from datasets import load_dataset
+from torch.utils.data import DataLoader
+from omegaconf import OmegaConf, DictConfig
+
+from train import val_epoch
+from data.batch import Batch
+from model import make_model
+from data.dataset import BinsToVelocityDataset
+from modules.encoderdecoder import subsequent_mask
+from modules.label_smoothing import LabelSmoothing
+
+
+def load_test_dataset(cfg: DictConfig):
```

Comment:
This logic is half-duplicated with `train.py` - can you try to generalise this? 

The HF dataset name shouldn't be hardcoded in here, but rather a part of the config. To deal with multiple possible split combinations, you can have this type of approach:

```python
def load_cached_dataset(cfg: DictConfig, split: str = "train"):
    dataset = load_dataset(dataset_name, split=split)
    
    config_string = json.dumps(cfg) + split
    ...
    return dataset
```

One issue with this: if you change the batch size, you'll get a different dataset cache, even though the data is exactly the same. A cleaner solution could be:

```python
def load_cached_dataset(dataset_cfg: DictConfig, split: str = "train"):
    ...
    BinsToVelocityDataset(
        dataset=dataset,
        n_dstart_bins=dataset_cfg.n_dstart_bins,
        n_velocity_bins=dataset_cfg.n_velocity_bins,
        n_duration_bins=dataset_cfg.n_duration_bins,
        sequence_len=dataset_cfg.sequence_size,
    )
```

Reply:
I do not fully understand how can I get a different dataset cache? I am passing only the 'dataset' section of train_cfg, so 'bins' and 'sequence_len', so changing train.batch_size should not be an issue I think?

Reply:
That's right, you can ignore the last part of my comment 🙈